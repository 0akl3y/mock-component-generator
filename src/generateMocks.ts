/* eslint-disable @typescript-eslint/no-explicit-any */
import { transform } from '@babel/core'
import generate from '@babel/generator'
import * as parser from '@babel/parser'
import traverse, { NodePath, Visitor } from '@babel/traverse'
import * as t from '@babel/types'

export interface MockGeneratorOptions {
  keepImports?: boolean
  keepTSTypes?: boolean
}

export const generateMock = (code: string, options?: MockGeneratorOptions) => {
  //strip typescript

  const transformedCode = options?.keepTSTypes
    ? code
    : transform(code, {
        plugins: [['@babel/plugin-transform-typescript', { isTSX: true }]],
        parserOpts: {
          sourceType: 'module',
          plugins: ['typescript', 'jsx'],
        },
      })?.code

  if (!transformedCode) {
    return
  }

  const ast = parser.parse(transformedCode, {
    sourceType: 'module',
    plugins: ['typescript', 'jsx'],
  })

  const mockFunctionBlockHelper = (functionName: string, params: any[]) =>
    t.variableDeclaration('const', [
      t.variableDeclarator(
        t.identifier(functionName),
        t.arrowFunctionExpression(
          params,
          t.blockStatement([
            t.returnStatement(
              t.callExpression(
                t.memberExpression(
                  t.identifier('React'),
                  t.identifier('createElement')
                ),
                [t.stringLiteral(functionName), ...params]
              )
            ),
          ]),
          false
        )
      ),
    ])

  const exportDefaultFunctionHelper = (functionName = '', params: any[]) =>
    t.exportDefaultDeclaration(
      t.arrowFunctionExpression(
        params,
        t.blockStatement([
          t.returnStatement(
            t.callExpression(
              t.memberExpression(
                t.identifier('React'),
                t.identifier('createElement')
              ),
              [t.stringLiteral(functionName), ...params]
            )
          ),
        ]),
        false
      )
    )

  const jestFn = t.callExpression(
    t.memberExpression(t.identifier('jest'), t.identifier('fn')),
    []
  )

  const jestFunctionBlockHelper = (functionName: string, params: any[]) =>
    t.variableDeclaration('const', [
      t.variableDeclarator(
        t.identifier(functionName),
        t.arrowFunctionExpression(params, jestFn)
      ),
    ])

  const hasExportDeclaration = (path: NodePath<any>) =>
    Boolean(
      path.findParent(
        (path) => path.isExportDeclaration() || path.isExportDefaultSpecifier()
      )
    )

  const isReactElementCreator = (expression: unknown) => {
    const callee = (expression as t.CallExpression)
      ?.callee as t.MemberExpression

    if (((callee as unknown) as t.Identifier)?.name === 'React.createElement') {
      return true
    }

    return (
      (callee?.object as t.Identifier)?.name === 'React' &&
      (callee?.property as t.Identifier)?.name === 'createElement'
    )
  }

  const isReactComponentClass = (node: unknown) => {
    const classNode = node as t.ClassDeclaration
    const className = classNode?.id?.name
    const superClass = classNode?.superClass as t.MemberExpression
    return (
      (superClass?.object as t.Identifier)?.name === 'React' &&
      ((superClass?.property as t.Identifier)?.name === 'PureComponent' ||
        (superClass?.property as t.Identifier)?.name === 'Component')
    )
  }

  const replaceDefaultExportReference = (args: {
    defaultExportName: string
  }): Visitor => ({
    //handle functional components

    ClassDeclaration(path: NodePath<t.ClassDeclaration>) {
      if (isReactComponentClass(path.node)) {
        const className = (path.node as t.ClassDeclaration)?.id?.name
        const mock = exportDefaultFunctionHelper(className ?? '', [
          t.identifier('props'),
        ])
        path?.replaceWith(mock)
      }
    },
    JSXElement(path: NodePath<t.JSXElement>) {
      const functionParent = path.getFunctionParent()

      if (functionParent) {
        const params = functionParent?.node?.params as any[]
        const functionPath = path.findParent(
          (path) => path.isVariableDeclarator() || path.isFunctionDeclaration()
        ) as
          | NodePath<t.VariableDeclarator>
          | NodePath<t.FunctionDeclaration>
          | null
        const functionName = (functionPath?.node as any)?.id?.name
        if (functionName === args.defaultExportName) {
          const exportDefaultFunction = exportDefaultFunctionHelper(
            functionName,
            params
          )
          if (functionPath?.isVariableDeclarator()) {
            const arrowFunctionVariableDeclaration = functionPath.findParent(
              (path) => path.isVariableDeclaration()
            )
            arrowFunctionVariableDeclaration?.replaceWith(exportDefaultFunction)
          } else {
            functionPath?.replaceWith(exportDefaultFunction)
          }
          path.skip()
        }
      }
    },
  })

  const handleCallExpressionAsDeclarationVistor: Visitor = {
    CallExpression(path: NodePath<t.CallExpression>) {
      const identifiers = path?.node?.arguments.filter((arg) =>
        t.isIdentifier(arg)
      ) as t.Identifier[]

      identifiers.forEach((identifier) => {
        const parentPath = path.findParent((path) => {
          return path.isProgram()
        }) as NodePath<t.Program>
        parentPath?.traverse(
          replaceDefaultExportReference({
            defaultExportName: identifier?.name ?? '',
          })
        )
      })
    },
  }

  traverse(ast, {
    ExportDefaultDeclaration(path: NodePath<t.ExportDefaultDeclaration>) {
      const defaultExportName = (path.node.declaration as t.Identifier)?.name
      const defaultCallExpression = path.node.declaration as t.CallExpression

      if (defaultExportName) {
        const parentPath = path.findParent((path) => {
          return path.isProgram()
        }) as NodePath<t.Program>
        parentPath?.traverse(
          replaceDefaultExportReference({ defaultExportName })
        )
      }

      if (defaultCallExpression) {
        path?.traverse(handleCallExpressionAsDeclarationVistor)
      }
    },
  })

  traverse(ast, {
    // remove all imports

    ImportDeclaration(path) {
      if (!options?.keepImports) {
        path.remove()
      }
    },

    VariableDeclaration(path: NodePath<t.VariableDeclaration>) {
      if (!hasExportDeclaration(path)) {
        path.remove()
      }
    },

    FunctionDeclaration(path: NodePath<t.FunctionDeclaration>) {
      if (!hasExportDeclaration(path)) {
        path.remove()
      }
    },

    TaggedTemplateExpression(path: NodePath<t.TaggedTemplateExpression>) {
      const parent = path.findParent((p) => t.isVariableDeclaration(p))
      parent?.remove()
    },

    ExportDefaultDeclaration(path: NodePath<t.ExportDefaultDeclaration>) {
      const declaration = path.node.declaration
      if (!t.isClass(declaration) && !t.isFunction(declaration)) {
        path.remove()
      }
      if (isReactComponentClass(declaration)) {
        const className = (declaration as t.ClassDeclaration)?.id?.name
        const mock = exportDefaultFunctionHelper(className ?? '', [
          t.identifier('props'),
        ])
        path?.replaceWith(mock)
      }
    },

    ExportNamedDeclaration(path: NodePath<t.ExportNamedDeclaration>) {
      const classDeclaration = path.node.declaration as t.ClassDeclaration
      const className = classDeclaration?.id?.name
      if (isReactComponentClass(classDeclaration)) {
        const mockedFunction = mockFunctionBlockHelper(className ?? '', [
          t.identifier('props'),
        ])
        path?.replaceWith(t.exportNamedDeclaration(mockedFunction))
      }
    },

    JSXElement(path) {
      let parent: NodePath<t.Function> | NodePath<t.Class> | null
      // eslint-disable-next-line prefer-const
      parent = path.findParent((path) => path.isFunction()) as typeof parent
      if (parent?.isFunction()) {
        const params = parent.node?.params as any[]
        const functionName = (path.findParent(
          (path: NodePath) =>
            path.isVariableDeclarator() || path.isFunctionDeclaration()
        )?.node as any)?.id.name
        const mockedElement = t.callExpression(
          t.memberExpression(
            t.identifier('React'),
            t.identifier('createElement')
          ),
          [t.stringLiteral(functionName ?? ''), ...params]
        )
        path.replaceWith(mockedElement)
      }
    },
    BlockStatement(path: NodePath<t.BlockStatement>) {
      if (
        !path.node.body.some((elm) => {
          const argument = (elm as t.ReturnStatement)?.argument
          return t.isJSXElement(argument) || isReactElementCreator(argument)
        })
      ) {
        const functionParent = path.getFunctionParent()
        if (t.isArrowFunctionExpression(functionParent)) {
          path.replaceInline(jestFn)
        } else if (t.isFunctionDeclaration(functionParent)) {
          const name =
            (functionParent.node as t.FunctionDeclaration)?.id?.name ?? ''
          functionParent.replaceWith(jestFunctionBlockHelper(name, []))
        }
        path.skip()
      }
    },
    ArrowFunctionExpression(path: NodePath<t.ArrowFunctionExpression>) {
      const body = path.node.body

      if (
        t.isExpression(body) &&
        !t.isJSXElement(body) &&
        !isReactElementCreator(body)
      ) {
        path.replaceInline(jestFn)
      }
    },
  })

  const output = generate(
    ast,
    {
      comments: false,
    },
    code
  )

  output.code = `import React from 'react'\n${output.code}`
  return output.code
}
