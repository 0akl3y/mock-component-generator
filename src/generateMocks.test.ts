import { generateMock } from "./generateMocks";

const code = `
import Stuff from '@/some-stuff'
import Blah from '@/some-blah'
import { Blub } from '@/some-blub'

export const SomeComponent = (props) => {
  return <Row />
}

export function someFunction() {
  return () => 'foo'
}
`;

describe("generateMocks", () => {
  it.skip("replaces the import statement with only react", () => {
    const input = `
    import Stuff from '@/some-stuff'
    import Blah from '@/some-blah'
    import { Blub } from '@/some-blub'

    const someOtherCode;
    `;
    expect(generateMock(input)).toEqual(
      "import React from 'react'\nconst someOtherCode;"
    );
  });

  describe("components", () => {
    it.skip("mocks functional components of different sort", () => {
      const input = `
      export const same = 'the same procedure as every year'
      export const shortComponent = props => <SomeView />
      export const parans = (props: {id: string}) => (<SomeViewInParans />)
      export const braced = (props: {id: string}) => {return (<SomeBracedView />)}
      export const someFunction = props => 'Remains the same'      
      `;
      expect(generateMock(input)).toMatchSnapshot();
    });

    it.skip("removes non exported components", () => {
      const input = `
      const _internalComponent = props => <SomeInternalComponent />      
      const _privateFN = () => 'should not show'
      export const externalComponent = props => <SomeExternalComponent />            
      `;
      expect(generateMock(input)).toMatchSnapshot();
    });

    it.skip("mocks functional components", () => {
      const input = `
      import React from 'react'
      import {Row} from 'superuiguide'
      const someStuff
        
      const curly = (input) => ('braces')
      const someOtherFunction = (input) => {return input + 1}
      
      export const SomeComponent = (props) => {
        return <Row />
      }
      `;
      expect(generateMock(input)).toMatchSnapshot();
    });

    it.skip("mocks non component functions with jest.fn", () => {
      const input = `
      export const function1 = (args: string[]) => { return 'function 1'}
      export const function2 = args => 'function 2'
      export const function3 = (args:string) => ('function 3')
      `;
      expect(generateMock(input)).toMatchSnapshot();
    });

    it.skip("mocks functions", () => {
      const input = `
      export function fn1(args) {
         return 'fn1'
      }

      export function FnComponent(props: {foo: string}) {
        return <Row />
      }
      
      `;
      expect(generateMock(input)).toMatchSnapshot();
    });

    it("removes comments", () => {
      const input = `
      // This is some unneccessary comment
      /*
        Once upon a time... blah blah blah      
      */
      export const function2 = args => 'function 2'
      
      `;
      expect(generateMock(input)).toMatchSnapshot();
    });
  });
});
