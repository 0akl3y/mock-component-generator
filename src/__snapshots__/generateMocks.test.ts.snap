// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMocks components handles direct export defaults correctly 1`] = `
"import React from 'react'
export default (props => {
  return React.createElement(\\"\\", props);
});"
`;

exports[`generateMocks components handles export default correctly 1`] = `
"import React from 'react'
export default (props => {
  return React.createElement(\\"SomeComponent\\", props);
});"
`;

exports[`generateMocks components handles export wrapped in higher order function 1`] = `
"import React from 'react'
export default (props => {
  return React.createElement(\\"SomeComponent\\", props);
});"
`;

exports[`generateMocks components handles styled components 1`] = `
"import React from 'react'
export const SomeComponent = props => {
  return React.createElement(\\"SomeComponent\\", props);
};"
`;

exports[`generateMocks components mocks class component 1`] = `
"import React from 'react'
export const SomeClassComponent = props => {
  return React.createElement(\\"SomeClassComponent\\", props);
};"
`;

exports[`generateMocks components mocks default class component 1`] = `
"import React from 'react'
export default (props => {
  return React.createElement(\\"SomeClassComponent\\", props);
});"
`;

exports[`generateMocks components mocks functional components 1`] = `
"import React from 'react'
export const SomeComponent = props => {
  return React.createElement(\\"SomeComponent\\", props);
};"
`;

exports[`generateMocks components mocks functional components of different sort 1`] = `
"import React from 'react'
export const same = 'the same procedure as every year';
export const shortComponent = props => React.createElement(\\"shortComponent\\", props);
export const parans = props => React.createElement(\\"parans\\", props);
export const braced = props => {
  return React.createElement(\\"braced\\", props);
};
export const someFunction = jest.fn();"
`;

exports[`generateMocks components mocks functions 1`] = `
"import React from 'react'
export const fn1 = jest.fn();
export function FnComponent(props) {
  return React.createElement(\\"FnComponent\\", props);
}"
`;

exports[`generateMocks components mocks non component functions with jest.fn 1`] = `
"import React from 'react'
export const function1 = args => jest.fn();
export const function2 = jest.fn();
export const function3 = jest.fn();"
`;

exports[`generateMocks components mocks referenced default class component 1`] = `
"import React from 'react'
export default (props => {
  return React.createElement(\\"SomeClassComponent\\", props);
});"
`;

exports[`generateMocks components removes comments 1`] = `
"import React from 'react'
export const function2 = jest.fn();"
`;

exports[`generateMocks components removes non exported components 1`] = `
"import React from 'react'
export const externalComponent = props => React.createElement(\\"externalComponent\\", props);"
`;

exports[`generateMocks components removes typescript 1`] = `
"import React from 'react'
export const parans = props => React.createElement(\\"parans\\", props);"
`;
